src/FiniteVolumeMLMC.cpp

```diff
@@ -0,0 +1,153 @@
+ *
+ * Copyright (c) 2018, Lawrence Livermore National Security, LLC.
+ * Produced at the Lawrence Livermore National Laboratory.
+ * LLNL-CODE-745247. All Rights reserved. See file COPYRIGHT for details.
+ *
+ * This file is part of smoothG. For more information and source code
+ * availability, see https://www.github.com/llnl/smoothG.
+ *
+ * smoothG is free software; you can redistribute it and/or modify it under the
+ * terms of the GNU Lesser General Public License (as published by the Free
+ * Software Foundation) version 2.1 dated February 1999.
+ *
+ ***********************************************************************EHEADER*/
+
+/** @file
+
+    @brief Implements FiniteVolumeMLMC class
+*/
+
+#include "FiniteVolumeMLMC.hpp"
+
+namespace smoothg
+{
+
+FiniteVolumeMLMC::FiniteVolumeMLMC(MPI_Comm comm,
+                                   const mfem::SparseMatrix& vertex_edge,
+                                   const mfem::Vector& weight,
+                                   const mfem::Array<int>& global_partitioning,
+                                   const mfem::HypreParMatrix& edge_d_td,
+                                   const mfem::SparseMatrix& edge_boundary_att,
+                                   const mfem::Array<int>& ess_attr,
+                                   double spect_tol, int max_evects)
+    : Upscale(comm, vertex_edge.Height(), false),
+      weight_(weight),
+      edge_d_td_(edge_d_td),
+      edge_boundary_att_(edge_boundary_att),
+      ess_attr_(ess_attr)
+{
+    const bool dual_target = false;
+    const bool scaled_dual = false;
+    const bool energy_dual = false;
+
+    mfem::StopWatch chrono;
+    chrono.Start();
+
+    // Hypre may modify the original vertex_edge, which we seek to avoid
+    mfem::SparseMatrix ve_copy(vertex_edge);
+
+    mixed_laplacians_.emplace_back(ve_copy, weight, edge_d_td_,
+                                   MixedMatrix::DistributeWeight::False);
+
+    auto graph_topology = make_unique<GraphTopology>(ve_copy, edge_d_td_, global_partitioning,
+                                                     &edge_boundary_att_);
+
+    mbuilder_ = make_unique<CoefficientMBuilder>(*graph_topology);
+    coarsener_ = make_unique<SpectralAMG_MGL_Coarsener>(
+                     mixed_laplacians_[0], std::move(graph_topology),
+                     spect_tol, max_evects, dual_target, scaled_dual, energy_dual,
+                     *mbuilder_);
+    coarsener_->construct_coarse_subspace();
+
+    mixed_laplacians_.push_back(coarsener_->GetCoarse());
+
+    MakeCoarseSolver();
+
+    MakeCoarseVectors();
+
+    chrono.Stop();
+    setup_time_ += chrono.RealTime();
+}
+
+/// this implementation is sloppy
+void FiniteVolumeMLMC::RescaleFineCoefficient(const mfem::Vector& coeff)
+{
+    mfem::Vector temp(coeff);
+    for (int i = 0; i < coeff.Size(); ++i)
+        temp(i) = 1.0 / coeff(i);
+    GetFineMatrix().ScaleM(temp);
```

Comment:
Like the coarse level update, the fine level will need to keep data of the original M, and produce a new M whenever an element scaling coefficient is given. The current way will rescale the rescaled matrix. So in mlmc.cpp, if more than 2 samples are tested, the coarse and fine level pressure will get different solutions

Reply:
Good catch, this is now fixed.